test/irmin-pack/test_upgrade.ml

```diff
@@ -31,409 +31,622 @@ let hash_of_string hash =
-      let entry = CA.read_and_decode_entry_prefix ~off fm in
-      let len = CA.Entry_prefix.total_entry_length entry |> Option.get in
-      (entry.hash, len)
-
-    let dict_find (repo : S.repo) step =
-      let dict = repo.dict in
-      S.Dict.index dict step |> Option.get
-
-    let index_find (repo : S.repo) hash =
-      let index = S.File_manager.index repo.fm in
-      S.Index.find index hash |> Option.get
-
-    let suffix_find (repo : S.repo) off = decode_entry_prefix ~off repo.fm
-
-    let index_iter (repo : S.repo) f =
-      let index = S.File_manager.index repo.fm in
-      S.Index.iter f index
-  end
+let c1 =
+  {
+    h = h "7e8978094f6c72e711310a959a91e7652eec8185";
+    o = i 232;
+    l = 51;
+    k = `c;
+  }
 
-  exception Skip_the_rest_of_that_test
+let borphan' =
+  {
+    h = h "945bcf284cb6f4735eb8eb74553637b43fde996b";
+    o = i 283;
+    l = 30;
+    k = `b;
+  }
 
-  let fail_and_skip root error =
-    let* () =
-      Alcotest.check_raises_lwt "open empty/V2 store in RO"
-        (Irmin_pack_unix.Errors.Pack_error error) (fun () ->
-          let* repo = Store.v ~readonly:true ~fresh:false root in
-          Store.close repo)
-    in
-    raise Skip_the_rest_of_that_test
-
-  (* Only checking injection and not bijection. *)
-  let test_dict repo model =
-    Hashtbl.iter
-      (fun step expected ->
-        let got = Store.dict_find repo step in
-        let msg = "---- dict " ^ step in
-        Alcotest.(check int) msg expected got)
-      model.Model.dict
-
-  let test_index repo model =
-    Hashtbl.iter
-      (fun hash expected ->
-        let got = Store.index_find repo hash in
-        let msg = Fmt.str "---- M index %a" Store.pp_hash hash in
-        Alcotest.(check (triple int63 int kind)) msg expected got)
-      model.Model.index;
-    Store.index_iter repo (fun hash expected ->
-        let got = Hashtbl.find model.Model.index hash in
-        let msg = Fmt.str "---- S index %a" Store.pp_hash hash in
-        Alcotest.(check (triple int63 int kind)) msg expected got)
-
-  (* Only checking injection and not bijection. *)
-  let test_suffix repo model =
-    Hashtbl.iter
-      (fun off expected ->
-        let got = Store.suffix_find repo off in
-        let msg = Fmt.str "---- suffix %a" Int63.pp off in
-        Alcotest.(check (pair hash int)) msg expected got)
-      model.Model.suffix
-
-  type mode = From_v2 | From_v3 | From_scratch [@@deriving irmin ~pp]
-
-  type phase = Before_rw_open | Between_rw_open_and_write | After_rw_write
-  [@@deriving irmin ~pp]
+let b2 =
+  {
+    h = h "32f28ea03b1b20126629d2ca63fc6665b0bbb604";
+    o = i 313;
+    l = 24;
+    k = `b;
+  }
 
+let n2 =
+  {
+    h = h "bbca871beaebb1b556e498a8e1ccae7817f5f4ff";
+    o = i 337;
+    l = 34;
+    k = `n;
+  }
+
+let c2 =
+  {
+    h = h "9abd8707d56dfbd484eb148c6be13887cecc0d59";
+    o = i 371;
+    l = 51;
+    k = `c;
+  }
+
+let pack_entries =
+  [ n0; b1; borphan; c2; c1; b01; borphan'; n1; n2; n01; c0; b2 ]
```

Comment:
maybe rename `borphan0` and `borphan1` so that its clearer when they are added

Reply:
I'd like to keep the digits to designate the commit successor(s).

I've added comments to improve that